### PR TITLE
Suryansh/testcases latest1

### PR DESCRIPTION
**Pull request by suryansh-max**

### CODE REVIEW

File: frontend/BoardGUI.py [hunk 1]
```diff
@@ -33,26 +35,45 @@ def draw_piece(self, piece, position):
         else:
             self.game.screen.blit(self.white_piece_image, position)
 
+    # def get_cell_clicked(self):
```
Comment (EXC3ll3NTrhyTHM) [inline frontend/BoardGUI.py:38]: Lets remove the commented code. We can access it later through the version history if we want to add it back.

File: frontend/BoardGUI.py [hunk 2]
```diff
@@ -33,26 +35,45 @@ def draw_piece(self, piece, position):
...
         col, row = map(int, current_cell)
-        left_click = pg.mouse.get_pressed()[0]
-        if left_click:
-            # print(self.backend_game.get_board(row,col), self.count)
-            # self.count += 1
+        mouse_buttons = pg.mouse.get_pressed()  # Call this once and store the result
 
+        if mouse_buttons[0]:  # Left click
+            self.count += 1
             if not self.backend_game.placement_complete():
```
Comment (EXC3ll3NTrhyTHM) [inline frontend/BoardGUI.py:68]: I believe in the next sprint we should put this into the GameManager class so that the BoardGUI isnt controlling the turn switching. I think it would be okay if the GameManager class draws the pieces onto the board or we could store the turn in a variable that we send back to the BoardGUI to keep it separate. We can discuss this more later.

File: frontend/BoardGUI.py [hunk 3]
```diff
@@ -33,26 +35,45 @@ def draw_piece(self, piece, position):
...
-                flag = self.backend_game.piece_placement(row, col)
-                # print(x, "\n")
-                if flag == 1:
-                    self.draw_piece(self.backend_game.get_turn(),current_cell * self.cell_size)
+                x = self.backend_game.piece_placement(row, col)
```
Comment (EXC3ll3NTrhyTHM) [inline frontend/BoardGUI.py:69]: This should be called `is_piece_placed` or something to that effect instead of `x`.

Comment (EXC3ll3NTrhyTHM) [reply frontend/BoardGUI.py:69]: Ultimately though we shouldnt need this variable because `piece_placement` should raise an exception if its invalid.

File: tests/frontend/test_BoardGUI.py [hunk 4]
```diff
@@ -1,73 +1,177 @@
...
+    mock_smoothscale.return_value = mock_image
+
+    # Call the method with a mock path and resolution and assert the calls were made correctly
+    path = 'resources/board/board.png'
+    resolution = [100, 100]
+    image = board_gui.get_scaled_image(path, resolution)
+
+    mock_load.assert_called_once_with(path)
+    mock_smoothscale.assert_called_once_with(mock_image.convert_alpha(), resolution)
+    assert image == mock_image
```
Comment (EXC3ll3NTrhyTHM) [inline tests/frontend/test_BoardGUI.py:63]: The true test for get_scaled_image should be that the returned image is the size of the resolution provided to the function.

We would also want a test for a negative path. If image.get_size() is not equal to resoultion then scaled_image.get_size() doesnt equal image.get_size()

Also Im not sure if you need to mock out the image and function here. Since we have the actual image path we can use and we can use the pygame.transform.smoothscale I think we can just use the actual artifacts to run the test without any mocks.

File: tests/frontend/test_BoardGUI.py [hunk 5]
```diff
@@ -1,73 +1,177 @@
...
+    resolution = (200, 200)  # Example of a different resolution
+    scaled_image = board_gui.get_scaled_image(path, resolution)
+
+    # Get the original image size
+    original_image = pg.image.load(path)
+    original_size = original_image.get_size()
+
+    # Assert that the sizes do not match
+    if original_size != resolution:
+        assert scaled_image.get_size() != original_size
```
Comment (EXC3ll3NTrhyTHM) [inline tests/frontend/test_BoardGUI.py:79]: Should do something similar in the above one.

File: tests/frontend/test_BoardGUI.py [hunk 6]
```diff
@@ -1,73 +1,177 @@
...
+    # Assuming draw_piece raises a ValueError for invalid positions
+    with pytest.raises(ValueError, match="Invalid position"):
+        board_gui.draw_piece('black', invalid_position)
+
+    with pytest.raises(ValueError, match="Invalid position"):
+        board_gui.draw_piece('white', invalid_position)
+
+
+
+def test_draw_piece(board_gui_fixture):
```
Comment (EXC3ll3NTrhyTHM) [inline tests/frontend/test_BoardGUI.py:161]: Duplicate? line 116

File: tests/frontend/test_BoardGUI.py [hunk 7]
```diff
@@ -1,73 +1,177 @@
...
+    calls = screen.blit.call_args_list
+    for call in calls:
+        if call[0][0] == piece_image and call[0][1] == position:
+            return True
+    return False
+
+
+
+# This test case failed because logic in backend is not correct to throw error
+def test_draw_piece_with_invalid_position(board_gui_fixture):
```
Comment (EXC3ll3NTrhyTHM) [inline tests/frontend/test_BoardGUI.py:148]: I think if we cant get it updated by tuesday we should adjust this test to reflect the current code. So that its not reflected poorly on us in grading.

File: tests/frontend/test_GameGUI.py [hunk 8]
```diff
@@ -1,64 +1,58 @@
...
 
-    def test_run_game(self):
-        # Mock methods to prevent an infinite loop and resource loading
-        with patch.object(self.game, 'check_events') as mock_check_events, \
-             patch.object(self.game.board, 'build_board') as mock_build_board, \
-             patch('pygame.display.update') as mock_display_update:
-            mock_check_events.side_effect = [None, SystemExit]  # Stop after one iteration
-            with self.assertRaises(SystemExit):
-                self.game.run_game()
+def test_run_game_quit(game_fixture):
```
Comment (EXC3ll3NTrhyTHM) [inline tests/frontend/test_GameGUI.py:39]: This is the other one thats failing for me

Comment (EXC3ll3NTrhyTHM) [reply tests/frontend/test_GameGUI.py:39]:
```
    def draw_board(self):
>       self.game.screen.blit(self.board_image, (0, 0))
E       TypeError: argument 1 must be pygame.surface.Surface, not Mock
```

File: tests/frontend/test_GameGUI.py [hunk 9]
```diff
@@ -1,64 +1,58 @@
...
-test_check_events_no_quit: It ensures that no exit actions are taken when QUIT is not in the events.
 
-test_run_game_quit: Tests that the run_game() method calls build_board() once and exits gracefully upon a QUIT event.
+def test_check_events_exit(game_fixture):
+    game = game_fixture
 
-test_run_game: Verifies that get_cell_clicked() is called multiple times as expected during the game loop. It simulates a series of events (KEYDOWN and QUIT) to control the game loop flow.
-
-'''
-class TestGameGUI(unittest.TestCase):
```
Comment (EXC3ll3NTrhyTHM) [inline tests/frontend/test_GameGUI.py:18]: How come you chose to not have any of the tests be in a class? ALl of our other tests are in classes.

**Comment by EXC3ll3NTrhyTHM** [comment]
Both files should contain test class that all test functions are within to align with established coding practices.

File: tests/frontend/test_GameGUI.py (hunk 9 above)
Comment (raspberrymilkyway) [reply tests/frontend/test_GameGUI.py:18]: Seconding this, but because we have to list classes on the documentation -- the tests need to be in classes 😅
